core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -60,11 +61,14 @@
 import javax.security.auth.Subject;
 
 /**
- * A shared context that isolates all operations within a {@link FileSystem}. Usually, one user
- * only needs one instance of {@link FileSystemContext}.
+ * A context that isolates all operations within a {@link FileSystem}. Typically a user
+ * will only need more than one instance of a {@link FileSystemContext} if they wish to connect
```

Comment:
Right now it's possible for them to create a context and pass it to `FileSystem.Factory.get` but I think because it contains internal methods like `acquireMasterClientResource` and `acquireBlockMasterClientResource` we should probably discourage the creation of this object by users and simply force them to pass the subject/Configuration instead.

Reply:
One thing I do foresee as a possible issue is users who might create many instances of `FileSystem` which would each have separate underlying `FileSystemContext`s. Each of those contexts would have their own thread pools, connections, resources, etc.. even if the same arguments were passed to the `FileSystem.Factoy`

Reply:
The way HDFS handles this is by having an optional (on by default) `FileSystem` [cache](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L458). Our HDFS API is by far the most commonly used api, so we will usually get filesystem caching for free. 

If we want caching outside the HDFS API, we could do something similar when users call our FileSystem.get.